From ae4ca2d69c7d8fb9432256c2843e509761ccbaec Mon Sep 17 00:00:00 2001
From: Chen Caidy <chen@caidy.cc>
Date: Wed, 19 Jul 2023 12:14:53 +0000
Subject: [PATCH] net: gptp: fix announce message byte order

Linuxptp report UTC offset is 9472 seconds, is a byte order issue,
The right value is 37. Also fixed offset_scaled_log_var byte order.

Signed-off-by: Chen Caidy <chen@caidy.cc>
Upstream-Status: Backport [main: 0635e2690e71cb423bb55b66c30e0a6a6a98b836]
---
 subsys/net/l2/ethernet/gptp/gptp_messages.c | 10 ++++++----
 1 file changed, 6 insertions(+), 4 deletions(-)

diff --git a/subsys/net/l2/ethernet/gptp/gptp_messages.c b/subsys/net/l2/ethernet/gptp/gptp_messages.c
index cb6fcf5bf3..8fc1ffe2ea 100644
--- a/subsys/net/l2/ethernet/gptp/gptp_messages.c
+++ b/subsys/net/l2/ethernet/gptp/gptp_messages.c
@@ -535,7 +535,7 @@ struct net_pkt *gptp_prepare_announce(int port)
 	hdr->reserved1 = 0U;
 	hdr->reserved2 = 0U;
 
-	ann->cur_utc_offset = global_ds->current_utc_offset;
+	ann->cur_utc_offset = htons(global_ds->current_utc_offset);
 	ann->time_source = global_ds->time_source;
 
 	switch (GPTP_PORT_BMCA_DATA(port)->info_is) {
@@ -543,9 +543,11 @@ struct net_pkt *gptp_prepare_announce(int port)
 		ann->root_system_id.grand_master_prio1 = default_ds->priority1;
 		ann->root_system_id.grand_master_prio2 = default_ds->priority2;
 
-		memcpy(&ann->root_system_id.clk_quality,
-		       &default_ds->clk_quality,
-		       sizeof(struct gptp_clock_quality));
+		ann->root_system_id.clk_quality.clock_accuracy =
+			default_ds->clk_quality.clock_accuracy;
+		ann->root_system_id.clk_quality.clock_class = default_ds->clk_quality.clock_class;
+		ann->root_system_id.clk_quality.offset_scaled_log_var =
+			htons(default_ds->clk_quality.offset_scaled_log_var);
 
 		memcpy(&ann->root_system_id.grand_master_id,
 		       default_ds->clk_id,
-- 
2.25.1

